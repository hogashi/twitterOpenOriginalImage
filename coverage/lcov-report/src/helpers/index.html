<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for src/helpers</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> src/helpers
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">41.1% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>90/219</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">26.32% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>30/114</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">41.3% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>19/46</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">41.47% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>90/217</span>
      </div>
    </div>
    <p class="quiet">
      Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
    </p>
  </div>
  <div class='status-line low'></div>
<div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file low" data-value="ButtonSetter.ts"><a href="ButtonSetter.ts.html">ButtonSetter.ts</a></td>
	<td data-value="25" class="pic low"><div class="chart"><div class="cover-fill" style="width: 25%;"></div><div class="cover-empty" style="width:75%;"></div></div></td>
	<td data-value="25" class="pct low">25%</td>
	<td data-value="104" class="abs low">26/104</td>
	<td data-value="7.27" class="pct low">7.27%</td>
	<td data-value="55" class="abs low">4/55</td>
	<td data-value="33.33" class="pct low">33.33%</td>
	<td data-value="24" class="abs low">8/24</td>
	<td data-value="25.49" class="pct low">25.49%</td>
	<td data-value="102" class="abs low">26/102</td>
	</tr>

<tr>
	<td class="file low" data-value="ButtonSetterTweetDeck.ts"><a href="ButtonSetterTweetDeck.ts.html">ButtonSetterTweetDeck.ts</a></td>
	<td data-value="0" class="pic low"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="51" class="abs low">0/51</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="33" class="abs low">0/33</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="11" class="abs low">0/11</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="51" class="abs low">0/51</td>
	</tr>

<tr>
	<td class="file high" data-value="ButtonSetters.ts"><a href="ButtonSetters.ts.html">ButtonSetters.ts</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	</tr>

<tr>
	<td class="file high" data-value="Constants.ts"><a href="Constants.ts.html">Constants.ts</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="14" class="abs high">14/14</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="14" class="abs high">14/14</td>
	</tr>

<tr>
	<td class="file high" data-value="Utils.ts"><a href="Utils.ts.html">Utils.ts</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="47" class="abs high">47/47</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="26" class="abs high">26/26</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="11" class="abs high">11/11</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="47" class="abs high">47/47</td>
	</tr>

</tbody>
</table>
</div><div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Sun Jul 07 2019 20:49:52 GMT+0900 (GMT+09:00)
</div>
</div>
<script src="../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../sorter.js"></script>
<script src="../../block-navigation.js"></script>
</body>
</html>
