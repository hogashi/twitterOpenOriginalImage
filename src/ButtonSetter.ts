import { isFalse, Options, SHOW_ON_TIMELINE, SHOW_ON_TWEET_DETAIL } from './constants';
import { onOriginalButtonClick, printException, setStyle } from './utils';

export interface ButtonSetterType {
  setButtonOnTimeline: (currentOptions: Options) => void;
  setButtonOnTweetDetail: (currentOptions: Options) => void;
}

/**
 * twitter.comでボタンを設置するクラス
 */
export class ButtonSetter implements ButtonSetterType {
  // タイムラインにボタン表示
  public setButtonOnTimeline(currentOptions: Options): void {
    // 昔のビューの処理はしばらく残す
    // ref: https://github.com/hogashi/twitterOpenOriginalImage/issues/32#issuecomment-578510155
    if (document.querySelector('#react-root')) {
      this._setButtonOnReactLayoutTimeline(currentOptions);
      return;
    }
    this._setButtonOnTimeline(currentOptions);
  }

  // ツイート詳細にボタン表示
  public setButtonOnTweetDetail(currentOptions: Options): void {
    // 昔のビューの処理はしばらく残す
    // TODO: Reactレイアウトでも実装する必要がある？
    // ref: https://github.com/hogashi/twitterOpenOriginalImage/issues/32#issuecomment-578510155
    this._setButtonOnTweetDetail(currentOptions);
  }

  private setButton({
    className,
    getImgSrcs,
    target,
  }: {
    className: string;
    getImgSrcs: () => string[];
    target: HTMLElement;
  }): void {
    const style = {
      width: '70px',
      'font-size': '13px',
      color: this.getActionButtonColor(),
    };

    /* つくるDOMは以下 */
    /*
      <div className='ProfileTweet-action tooi-button-container'>
        <input
          className='tooi-button'
          style={style}
          type='button'
          value='Original'
          onClick={(e) => {
            onOriginalButtonClick(e, imgSrcs);
          }}
        />
      </div>
    */

    const button = document.createElement('input');
    button.className = 'tooi-button';
    setStyle(button, style);
    button.type = 'button';
    button.value = 'Original';
    button.addEventListener('click', (e) => {
      onOriginalButtonClick(e, getImgSrcs());
    });

    const container = document.createElement('div');
    container.classList.add('ProfileTweet-action', className);

    target.appendChild(container);
    container.appendChild(button);
  }

  private setReactLayoutButton({
    className,
    getImgSrcs,
    target,
  }: {
    className: string;
    getImgSrcs: () => string[];
    target: HTMLElement;
  }): void {
    const button = document.createElement('input');

    button.type = 'button';
    button.value = 'Original';
    const color = this.getReactLayoutActionButtonColor();
    setStyle(button, {
      'font-size': '13px',
      padding: '4px 8px',
      color,
      'background-color': 'rgba(0, 0, 0, 0)',
      border: `1px solid ${color}`,
      'border-radius': '3px',
      cursor: 'pointer',
    });
    button.addEventListener('click', (e) => {
      onOriginalButtonClick(e, getImgSrcs());
    });

    const container = document.createElement('div');
    // container.id = '' + tweet.id
    container.classList.add(className);
    setStyle(container, {
      display: 'flex',
      'margin-left': '20px',
      'flex-flow': 'column',
      'justify-content': 'center',
    });

    target.appendChild(container);
    container.appendChild(button);
  }

  private _setButtonOnTimeline(currentOptions: Options): void {
    // タイムラインにボタン表示する設定がされているときだけ実行する
    // - isTrue か 設定なし のとき ON
    // - isFalse のとき OFF
    if (!(currentOptions[SHOW_ON_TIMELINE] !== isFalse)) {
      return;
    }
    const tweets = document.getElementsByClassName('js-stream-tweet');
    if (!tweets.length) {
      return;
    }
    const className = 'tooi-button-container-timeline';
    // 各ツイートに対して
    Array.from(tweets).forEach((tweet) => {
      // 画像ツイートかつまだ処理を行っていないときのみ行う
      if (
        !(
          tweet.getElementsByClassName('AdaptiveMedia-container').length !== 0 &&
          tweet.getElementsByClassName('AdaptiveMedia-container')[0].getElementsByTagName('img').length !== 0
        ) ||
        tweet.getElementsByClassName(className).length !== 0
      ) {
        return;
      }
      // 操作ボタンの外側は様式にあわせる
      const actionList = tweet.querySelector<HTMLElement>('.ProfileTweet-actionList');
      if (!actionList) {
        printException('no target');
        return;
      }

      // 画像の親が取得できたら
      const mediaContainer = tweet.getElementsByClassName('AdaptiveMedia-container')[0] as HTMLElement;
      const getImgSrcs = (): string[] =>
        Array.from(mediaContainer.getElementsByClassName('AdaptiveMedia-photoContainer')).map(
          (element) => element.getElementsByTagName('img')[0].src,
        );
      this.setButton({
        className,
        getImgSrcs,
        target: actionList,
      });
    });
  }

  private _setButtonOnTweetDetail(currentOptions: Options): void {
    // ツイート詳細にボタン表示する設定がされているときだけ実行する
    // - isTrue か 設定なし のとき ON
    // - isFalse のとき OFF
    if (!(currentOptions[SHOW_ON_TWEET_DETAIL] !== isFalse)) {
      return;
    }
    const className = 'tooi-button-container-detail';
    if (
      !document
        .getElementsByClassName('permalink-tweet-container')[0]
        ?.getElementsByClassName('AdaptiveMedia-photoContainer')[0] ||
      document.getElementsByClassName(className).length !== 0
    ) {
      // ツイート詳細ページでない、または、メインツイートが画像ツイートでないとき
      // または、すでに処理を行ってあるとき
      // 何もしない
      return;
    }
    // Originalボタンの親の親となる枠
    const actionList = document.querySelector<HTMLElement>('.permalink-tweet-container .ProfileTweet-actionList');
    if (!actionList) {
      printException('no target');
      return;
    }

    // .AdaptiveMedia-photoContainer: 画像のエレメントからURLを取得する
    const getImgSrcs = (): string[] =>
      Array.from(
        document
          .getElementsByClassName('permalink-tweet-container')[0]
          .getElementsByClassName('AdaptiveMedia-photoContainer'),
      ).map((element) => element.getElementsByTagName('img')[0].src);
    this.setButton({
      className,
      getImgSrcs,
      target: actionList,
    });
  }

  private _setButtonOnReactLayoutTimeline(currentOptions: Options): void {
    // ツイート詳細にボタン表示する設定がされているときだけ実行する
    // - isTrue か 設定なし のとき ON
    // - isFalse のとき OFF
    if (!(currentOptions[SHOW_ON_TIMELINE] !== isFalse)) {
      return;
    }
    const className = 'tooi-button-container-react-timeline';
    const tweets = Array.from(document.querySelectorAll('#react-root main section article'));
    if (!tweets.length) {
      return;
    }
    // 各ツイートに対して
    tweets.forEach((tweet) => {
      // 画像ツイート かつ 画像が1枚でもある かつ まだ処理を行っていないときのみ実行
      const tweetATags = Array.from(tweet.querySelectorAll('a')).filter((aTag) =>
        /\/status\/[0-9]+\/photo\//.test(aTag.href),
      );
      if (
        tweetATags.length === 0 ||
        tweetATags.every((aTag) => !aTag.querySelector('img')) ||
        tweet.getElementsByClassName(className)[0]
      ) {
        return;
      }
      // ボタンを設置
      // 操作ボタンの外側は様式にあわせる
      const target = tweet.querySelector<HTMLElement>('div[role="group"]');
      if (!target) {
        printException('no target');
        return;
      }

      const getImgSrcs = (): string[] => {
        const tweetImgs = tweetATags.map((aTag) => aTag.querySelector('img'));
        return tweetImgs
          .map((img) =>
            // filter で string[] にするためにここで string[] にする……
            img ? img.src : '',
          )
          .filter((src) => src !== '');
      };

      this.setReactLayoutButton({
        className,
        getImgSrcs,
        target,
      });
    });
  }

  private getActionButtonColor(): string {
    // コントラスト比4.5(chromeの推奨する最低ライン)の色
    const contrastLimitColor = '#697b8c';

    const actionButton = document.querySelector('.ProfileTweet-actionButton') as HTMLElement;
    if (!actionButton?.style) {
      return contrastLimitColor;
    }

    const buttonColor = window.getComputedStyle(actionButton).color;
    if (buttonColor && buttonColor.length > 0) {
      return buttonColor;
    }
    return contrastLimitColor;
  }

  private getReactLayoutActionButtonColor(): string {
    // 文字色
    // 初期値: コントラスト比4.5(chromeの推奨する最低ライン)の色
    let color = '#697b8c';
    // ツイートアクション(返信とか)のボタンのクラス(夜間モードか否かでクラス名が違う)
    const actionButton = document.querySelector<HTMLElement>('div[role="group"] div[role="button"]');
    if (actionButton?.children[0] && (actionButton.children[0] as HTMLElement).style) {
      const buttonColor = window.getComputedStyle(actionButton.children[0]).color;
      if (buttonColor && buttonColor.length > 0) {
        color = buttonColor;
      }
    }

    return color;
  }
}
