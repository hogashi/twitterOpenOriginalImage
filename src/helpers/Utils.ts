import { GET_LOCAL_STORAGE, Options } from './Constants';
import { MessageRequest, MessageResponse } from '../background';

// エラーメッセージの表示(予期せぬ状況の確認)
export const printException = (tooiException: string) => {
  console.log('tooi: ' + tooiException + ' at: ' + window.location.href);
};

// 画像URLを https～?format=〜&name=orig に揃える
export const formatUrl = (imgUrl: string) => {
  if (!imgUrl) {
    return null;
  }

  if (!/https:\/\/pbs\.twimg\.com\/media/.test(imgUrl)) {
    // twitterの画像URLでないときそのまま返す
    return imgUrl;
  }

  const url = new URL(imgUrl);
  const searchSet: { [key: string]: string } = {
    format: 'jpg',  // 拡張子が無い場合はjpgにフォールバック
  };
  url.search
    .slice(1)
    .split('&')
    .forEach(set => {
      const [key, value] = set.split('=');
      searchSet[key] = value;
    });

  const matched = url.pathname.match(
    /^(.*?)(?:|\.([^.:]+))(?:|:[a-z]+)$/
  );
  const pathname = matched[1];
  const extension = matched[2];
  // 2.1.11時点ではクエリパラメータを使うのはTweetDeckのみ
  // TweetDeckのURLでは拡張子を優先する
  // ref: https://hogashi.hatenablog.com/entry/2018/08/15/042044
  return `${url.protocol}//${url.host}${pathname}?format=${extension || searchSet.format}&name=orig`;
};

// 画像を開く
export const openImages = (imgSrcs: string[]) => {
  Array.from(imgSrcs)
    .reverse()  // 逆順に開くことで右側のタブから読める
    .forEach((imgSrc) => {
      // if 画像URLが取得できたなら
      const url = formatUrl(imgSrc);
      if (url) {
        window.open(url);
      } else {
        printException('no image url');
      }
    });
};

// 設定項目更新
export const updateOptions = (options: Options) => {
  console.log('update options: ', options) // debug
  return Promise.all(
    Object.keys(options).map(key =>
      new Promise((resolve, reject) => {
        const request: MessageRequest = {
          method: GET_LOCAL_STORAGE,
          key,
        };
        const callback = (response: MessageResponse) => {
          if (response && response.data) {
            options[key] = response.data;
            resolve();
          } else {
            reject();
          }
        };
        chrome.runtime.sendMessage(request, callback);
      })
    )
  ).then(() => {
    console.log('update options: ', options) // debug
  });
};
