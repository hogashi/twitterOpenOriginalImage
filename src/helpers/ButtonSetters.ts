import ButtonSetter from './ButtonSetter';
import ButtonSetterTweetDeck from './ButtonSetterTweetDeck';
import { HOST_TWITTER_COM, HOST_TWEETDECK_TWITTER_COM } from './Constants';

interface ButtonSettersType {
  [key: string]: ButtonSetter | ButtonSetterTweetDeck;
}

// ボタンを設置するクラスのまとめ
const ButtonSetters: ButtonSettersType = {};

ButtonSetters[HOST_TWITTER_COM] = new ButtonSetter();
ButtonSetters[HOST_TWEETDECK_TWITTER_COM] = new ButtonSetterTweetDeck();

export default ButtonSetters;
