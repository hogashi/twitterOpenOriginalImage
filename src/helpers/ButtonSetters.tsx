import ButtonSetter from './ButtonSetter';
import ButtonSetterTweetDeck from './ButtonSetterTweetDeck';
import * as Constants from './Constants';

interface ButtonSettersType {
  [key: string]: ButtonSetter | ButtonSetterTweetDeck;
}

// ボタンを設置するクラスのまとめ
const ButtonSetters: ButtonSettersType = {};

ButtonSetters[Constants.HOST_TWITTER_COM] = new ButtonSetter;
ButtonSetters[Constants.HOST_TWEETDECK_TWITTER_COM] = new ButtonSetterTweetDeck;

export default ButtonSetters;
