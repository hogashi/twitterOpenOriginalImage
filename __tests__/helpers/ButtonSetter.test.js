import ButtonSetter from '../../src/helpers/ButtonSetter';
import {
  INITIAL_OPTIONS,
  SHOW_ON_TIMELINE,
  isTrue,
  isFalse,
} from '../../src/helpers/Constants';
import { openImages } from '../../src/helpers/Utils';

jest.mock('../../src/helpers/Utils');

describe('ButtonSetter', () => {
  describe('setButtonOnTimeline', () => {
    it('従来のレイアウト(not React)', () => {
      const buttonSetter = new ButtonSetter();
      buttonSetter._setButtonOnTimeline = jest.fn();
      buttonSetter._setButtonOnReactLayoutTimeline = jest.fn();

      buttonSetter.setButtonOnTimeline(INITIAL_OPTIONS);

      expect(buttonSetter._setButtonOnTimeline).toHaveBeenCalledTimes(1);
      expect(buttonSetter._setButtonOnTimeline.mock.calls[0][0]).toStrictEqual(
        INITIAL_OPTIONS
      );

      expect(
        buttonSetter._setButtonOnReactLayoutTimeline
      ).not.toHaveBeenCalled();
    });

    it('新しいレイアウト(React)', () => {
      const buttonSetter = new ButtonSetter();
      buttonSetter._setButtonOnTimeline = jest.fn();
      buttonSetter._setButtonOnReactLayoutTimeline = jest.fn();

      const root = document.createElement('div');
      root.setAttribute('id', 'react-root');
      document.querySelector('body').appendChild(root);
      buttonSetter.setButtonOnTimeline(INITIAL_OPTIONS);

      expect(buttonSetter._setButtonOnTimeline).not.toHaveBeenCalled();

      expect(
        buttonSetter._setButtonOnReactLayoutTimeline
      ).toHaveBeenCalledTimes(1);
      expect(
        buttonSetter._setButtonOnReactLayoutTimeline.mock.calls[0][0]
      ).toStrictEqual(INITIAL_OPTIONS);
    });
  });

  describe('setButtonOnTweetDetail', () => {
    it('詳細ツイートでもボタン置く(従来レイアウトのみ)', () => {
      const buttonSetter = new ButtonSetter();
      buttonSetter._setButtonOnTweetDetail = jest.fn();

      buttonSetter.setButtonOnTweetDetail(INITIAL_OPTIONS);
      expect(buttonSetter._setButtonOnTweetDetail).toHaveBeenCalledTimes(1);
      expect(
        buttonSetter._setButtonOnTweetDetail.mock.calls[0][0]
      ).toStrictEqual(INITIAL_OPTIONS);
    });
  });

  describe('onClick', () => {
    it('イベント中断して画像開くメソッド呼ばれる', () => {
      const event = {
        preventDefault: jest.fn(),
        stopPropagation: jest.fn(),
      };
      const imgSrcs = ['src1', 'src2'];

      const buttonSetter = new ButtonSetter();
      buttonSetter.onClick(event, imgSrcs);

      expect(event.preventDefault).toHaveBeenCalledTimes(1);
      expect(event.stopPropagation).toHaveBeenCalledTimes(1);

      expect(openImages).toHaveBeenCalledTimes(1);
      expect(openImages.mock.calls[0][0]).toBe(imgSrcs);
    });
  });

  describe('setStyle', () => {
    it('objectからエレメントにstyle属性をつける', () => {
      const element = document.createElement('div');
      const style = {
        display: 'block',
        'background-color': 'green',
      };

      // 最初はセットされていない
      expect(element.style.display).toBe('');
      expect(element.style.backgroundColor).toBe('');

      const buttonSetter = new ButtonSetter();
      buttonSetter.setStyle(element, style);

      expect(element.style.display).toBe('block');
      expect(element.style.backgroundColor).toBe('green');
    });
  });

  describe('setButton', () => {
    const className = 'hogeclass';
    const imgSrcs = ['src1', 'src2'];
    const getImgSrcs = () => imgSrcs;
    const target = document.createElement('div');

    const buttonSetter = new ButtonSetter();
    buttonSetter.setButton({ className, getImgSrcs, target });

    it('ボタン設置される', () => {
      expect(target.innerHTML).toMatchSnapshot();
    });

    it('ボタン押すとonClick呼ばれる', () => {
      buttonSetter.onClick = jest.fn();
      const button = target.querySelector('input');
      button.click();
      expect(buttonSetter.onClick).toHaveBeenCalledTimes(1);
      expect(buttonSetter.onClick.mock.calls[0][0]).toBeInstanceOf(MouseEvent);
      expect(buttonSetter.onClick.mock.calls[0][1]).toStrictEqual(imgSrcs);
    });
  });

  describe('setReactLayoutButton', () => {
    const className = 'hogeclass';
    const imgSrcs = ['src1', 'src2'];
    const getImgSrcs = () => imgSrcs;
    const target = document.createElement('div');

    const buttonSetter = new ButtonSetter();
    buttonSetter.setReactLayoutButton({ className, getImgSrcs, target });

    it('ボタン設置される', () => {
      expect(target.innerHTML).toMatchSnapshot();
    });

    it('ボタン押すとonClick呼ばれる', () => {
      buttonSetter.onClick = jest.fn();
      const button = target.querySelector('input');
      button.click();
      expect(buttonSetter.onClick).toHaveBeenCalledTimes(1);
      expect(buttonSetter.onClick.mock.calls[0][0]).toBeInstanceOf(MouseEvent);
      expect(buttonSetter.onClick.mock.calls[0][1]).toStrictEqual(imgSrcs);
    });
  });

  describe('_setButtonOnTimeline', () => {
    /**
     * @param {string[]} imgSrcs
     * @param {HTMLElement[]} extraElements
     * @param {boolean} hasButton
     */
    const makeTweet = (imgSrcs, extraElements = [], hasButton = false) => {
      const root = document.createElement('div');
      root.classList.add('js-stream-tweet');

      const media = document.createElement('div');
      media.classList.add('AdaptiveMedia-container');

      const footer = document.createElement('footer');
      footer.classList.add('ProfileTweet-actionList');
      if (hasButton) {
        const button = document.createElement('div');
        button.classList.add('tooi-button-container-timeline');
        footer.appendChild(button);
      }

      imgSrcs.forEach(src => {
        const div = document.createElement('div');
        div.classList.add('AdaptiveMedia-photoContainer');
        const img = document.createElement('img');
        img.src = src;
        div.appendChild(img);
        media.appendChild(div);
      });

      extraElements.forEach(element => media.appendChild(element));

      root.appendChild(media);
      root.appendChild(footer);
      document.body.appendChild(root);
    };

    beforeEach(() => {
      document.body.innerHTML = '';
    });

    it('画像1枚ツイート1つにボタンつけようとする', () => {
      const imgSrcs = ['https://g.co/img1'];
      makeTweet(imgSrcs);

      const buttonSetter = new ButtonSetter();
      buttonSetter.setButton = jest.fn();

      const options = INITIAL_OPTIONS;
      options[SHOW_ON_TIMELINE] = isTrue;
      buttonSetter.setButtonOnTimeline(options);

      expect(buttonSetter.setButton).toHaveBeenCalledTimes(1);
      expect(buttonSetter.setButton.mock.calls[0][0].className).toStrictEqual(
        'tooi-button-container-timeline'
      );
      expect(
        buttonSetter.setButton.mock.calls[0][0].getImgSrcs()
      ).toMatchObject(imgSrcs);
      expect(
        buttonSetter.setButton.mock.calls[0][0].target.tagName
      ).toStrictEqual('FOOTER');
    });

    it('画像1枚ツイート3つにボタンつけようとする', () => {
      const imgSrcsSet = [
        ['https://g.co/img1'],
        ['https://g.co/img2'],
        ['https://g.co/img3'],
      ];
      imgSrcsSet.forEach(imgSrcs => {
        makeTweet(imgSrcs);
      });

      const buttonSetter = new ButtonSetter();
      buttonSetter.setButton = jest.fn();

      const options = INITIAL_OPTIONS;
      options[SHOW_ON_TIMELINE] = isTrue;
      buttonSetter.setButtonOnTimeline(options);

      expect(buttonSetter.setButton).toHaveBeenCalledTimes(3);
      imgSrcsSet.forEach((imgSrcs, index) => {
        expect(
          buttonSetter.setButton.mock.calls[index][0].className
        ).toStrictEqual('tooi-button-container-timeline');
        expect(
          buttonSetter.setButton.mock.calls[index][0].getImgSrcs()
        ).toMatchObject(imgSrcs);
        expect(
          buttonSetter.setButton.mock.calls[index][0].target.tagName
        ).toStrictEqual('FOOTER');
      });
    });

    it('画像4枚ツイート3つにボタンつけようとする', () => {
      const imgSrcsSet = [
        [
          'https://g.co/img11',
          'https://g.co/img12',
          'https://g.co/img13',
          'https://g.co/img14',
        ],
        [
          'https://g.co/img21',
          'https://g.co/img22',
          'https://g.co/img23',
          'https://g.co/img24',
        ],
        [
          'https://g.co/img31',
          'https://g.co/img32',
          'https://g.co/img33',
          'https://g.co/img34',
        ],
      ];
      imgSrcsSet.forEach(imgSrcs => {
        makeTweet(imgSrcs);
      });

      const buttonSetter = new ButtonSetter();
      buttonSetter.setButton = jest.fn();

      const options = INITIAL_OPTIONS;
      options[SHOW_ON_TIMELINE] = isTrue;
      buttonSetter.setButtonOnTimeline(options);

      expect(buttonSetter.setButton).toHaveBeenCalledTimes(3);
      imgSrcsSet.forEach((imgSrcs, index) => {
        expect(
          buttonSetter.setButton.mock.calls[index][0].className
        ).toStrictEqual('tooi-button-container-timeline');
        expect(
          buttonSetter.setButton.mock.calls[index][0].getImgSrcs()
        ).toMatchObject(imgSrcs);
        expect(
          buttonSetter.setButton.mock.calls[index][0].target.tagName
        ).toStrictEqual('FOOTER');
      });
    });

    it('動画/GIFのツイート1つにボタンつけない', () => {
      const imgSrcs = [];
      makeTweet(imgSrcs, [document.createElement('video')]);

      const buttonSetter = new ButtonSetter();
      buttonSetter.setButton = jest.fn();

      const options = INITIAL_OPTIONS;
      options[SHOW_ON_TIMELINE] = isTrue;
      buttonSetter.setButtonOnTimeline(options);

      expect(buttonSetter.setButton).not.toHaveBeenCalled();
    });

    it('画像1枚ツイート1つでも,もうボタンあったらボタンつけない', () => {
      const imgSrcs = ['https://g.co/img1'];
      makeTweet(imgSrcs, [], true);

      const buttonSetter = new ButtonSetter();
      buttonSetter.setButton = jest.fn();

      const options = INITIAL_OPTIONS;
      options[SHOW_ON_TIMELINE] = isTrue;
      buttonSetter.setButtonOnTimeline(options);

      expect(buttonSetter.setButton).not.toHaveBeenCalled();
    });

    it('画像ツイート1つあっても画像なかったらボタンつけない', () => {
      const imgSrcs = [];
      makeTweet(imgSrcs);

      const buttonSetter = new ButtonSetter();
      buttonSetter.setButton = jest.fn();

      const options = INITIAL_OPTIONS;
      options[SHOW_ON_TIMELINE] = isTrue;
      buttonSetter.setButtonOnTimeline(options);

      expect(buttonSetter.setButton).not.toHaveBeenCalled();
    });

    it('ツイート1つあっても画像ツイートじゃなかったらボタンつけない', () => {
      const imgSrcs = [];
      makeTweet(imgSrcs);

      const container = document.querySelector('.AdaptiveMedia-container');
      container.parentNode.removeChild(container);

      const buttonSetter = new ButtonSetter();
      buttonSetter.setButton = jest.fn();

      const options = INITIAL_OPTIONS;
      options[SHOW_ON_TIMELINE] = isTrue;
      buttonSetter.setButtonOnTimeline(options);

      expect(buttonSetter.setButton).not.toHaveBeenCalled();
    });

    it('footerがなかったらボタンつけない', () => {
      const imgSrcs = ['https://g.co/img1'];
      makeTweet(imgSrcs);

      const footer = document.querySelector('footer');
      footer.parentNode.removeChild(footer);

      const buttonSetter = new ButtonSetter();
      buttonSetter.setButton = jest.fn();

      const options = INITIAL_OPTIONS;
      options[SHOW_ON_TIMELINE] = isTrue;
      buttonSetter.setButtonOnTimeline(options);

      expect(buttonSetter.setButton).not.toHaveBeenCalled();
    });

    it('画像ツイートなかったら何もしない', () => {
      const buttonSetter = new ButtonSetter();
      buttonSetter.setButton = jest.fn();

      const options = INITIAL_OPTIONS;
      options[SHOW_ON_TIMELINE] = isTrue;
      buttonSetter.setButtonOnTimeline(options);

      expect(buttonSetter.setButton).not.toHaveBeenCalled();
    });

    it('設定がOFFなら何もしない', () => {
      const buttonSetter = new ButtonSetter();
      buttonSetter.setButton = jest.fn();

      const options = INITIAL_OPTIONS;
      options[SHOW_ON_TIMELINE] = isFalse;
      buttonSetter.setButtonOnTimeline(options);

      expect(buttonSetter.setButton).not.toHaveBeenCalled();
    });
  });
});
