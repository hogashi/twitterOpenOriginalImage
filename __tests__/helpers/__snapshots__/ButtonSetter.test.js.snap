// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonSetters setButton ボタン設置される 1`] = `"<div class=\\"ProfileTweet-action hogeclass\\"><input class=\\"tooi-button\\" style=\\"width: 70px; font-size: 13px; color: #697b8c\\" type=\\"button\\" value=\\"Original\\"></div>"`;
